examples/nextjs-flow/app/hooks/newPartials.ts

```diff
@@ -0,0 +1,369 @@
+      if (r.type === 'punctuation') {
+        return {
+          ...word,
+          punctuation: true,
+          eos: r.is_eos ?? false,
+          attachesTo: r.attaches_to,
+        };
+      }
+      return { ...word, punctuation: false };
+    });
+  }
+
+  /**
+   * Generates the current state of the transcript
+   * Groups words and responses by speaker and chronological order
+   */
+  getTranscriptGroups(): TranscriptGroup[] {
+    const sentences = this.getSentences();
+    const sentencesAndResponses = [...sentences, ...this.agentResponses];
+
+    // Sort everything by start time
+    sentencesAndResponses.sort((a, b) => {
+      const startTimeA = 'agent' in a ? a.startTime : a[0].startTime;
+      const startTimeB = 'agent' in b ? b.startTime : b[0].startTime;
+      return startTimeA - startTimeB;
+    });
+
+    const allSorted = sentencesAndResponses.flat();
+    const groups: TranscriptGroup[] = [];
+    let currentGroup: TranscriptGroup | undefined;
+
+    // Group consecutive items from the same speaker together
+    for (const wordOrResponse of allSorted) {
+      if (!currentGroup) {
+        currentGroup =
+          'agent' in wordOrResponse
+            ? { type: 'agent', data: [wordOrResponse] }
+            : {
+                type: 'speaker',
+                data: [wordOrResponse],
+                speaker: wordOrResponse.speaker,
+              };
+        continue;
+      }
+
+      if (currentGroup.type === 'agent' && 'agent' in wordOrResponse) {
+        currentGroup.data.push(wordOrResponse);
+      } else if (
+        currentGroup.type === 'speaker' &&
+        !('agent' in wordOrResponse) &&
+        wordOrResponse.speaker === currentGroup.speaker
+      ) {
+        currentGroup.data.push(wordOrResponse);
+      } else {
+        groups.push(currentGroup);
+        currentGroup =
+          'agent' in wordOrResponse
+            ? { type: 'agent', data: [wordOrResponse] }
+            : {
+                type: 'speaker',
+                data: [wordOrResponse],
+                speaker: wordOrResponse.speaker,
+              };
+      }
+    }
+
+    if (currentGroup) {
+      groups.push(currentGroup);
+    }
+
+    return groups;
+  }
+
+  /**
+   * Groups words into sentences based on punctuation and end-of-sentence markers
+   */
+  private getSentences(): Word[][] {
+    const sentences: Word[][] = [];
+    let currentSentence: Word[] = [];
+
+    for (const word of [...this.finals, ...this.partials]) {
+      currentSentence.push(word);
+      if (word.punctuation && word.eos) {
+        sentences.push(currentSentence);
+        currentSentence = [];
+      }
+    }
+
+    if (currentSentence.length) {
+      sentences.push(currentSentence);
+    }
+
+    return sentences;
+  }
+
+  /**
+   * Notifies listeners of updates to the transcript
+   * Dispatches a CustomEvent with the current transcript groups
+   */
+  private notifyUpdate() {
+    this.dispatchEvent(
+      new CustomEvent('update', {
```

Comment:
Nitpick, but I often find creating a custom class extending `Event` is easier to work with than `CustomEvent`s, since then you can express the type of whatever data they contain.